### PR TITLE
Fixing CopyFrom method in OdataMessageReader and OdataMessageWriter to include Arraypool

### PR DESCRIPTION
**Pull request by Sreejithpin**
### Issues

There was an issue that  OData.Core ICharArrayPool is not hooked up properly, which was causing lot of Large Object Heap allocations taking up lot of memory

### Description

- Added Arrayppol copying in ODataMessage reader and writer settings
- Added test cases for checking the same
- Added generic test method to compare 2 objects (deep copy)

### Checklist (Uncheck if it is not completed)

- [ ] *Test cases added*
- [ ] *Build and test with one-click build and test script passed*

### Additional work necessary
N/A

### CODE REVIEW

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ODataMessageReaderSettingsTests.cs [hunk 1]
```diff
@@ -132,7 +132,7 @@ public void ODataMessageReaderCopyConstructorTest()
             // Compare original and settings created from copy constructor after setting some values
             settings.BaseUri = new Uri("http://www.odata.org");
             settings.EnableCharactersCheck = true;
-            copyOfSettings = settings.Clone();
+            copyOfSettings = settings.Clone();
```
Comment (KanishManuja-MS) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/ODataMessageReaderSettingsTests.cs]: Unintended change.

File: test/FunctionalTests/Microsoft.OData.Core.Tests/TestCharArrayPool.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        {
+            return new char[minSize];
+        }
+
+        public void Return(char[] array)
+        {
+
+        }
+    }
+
```
Comment (KanishManuja-MS) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/TestCharArrayPool.cs]: Remove this line.

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ValidationHelper.cs [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+                            if (element1 == null && element2 == null)
+                            {
+                                continue;
+                            }
+
+                            if (element1 == null || element2 == null)
+                            {
+                                diff.Add(string.Format("Value of items in Collection {0} does not match.", prop.Name));
+
```
Comment (KanishManuja-MS) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/ValidationHelper.cs]: extra line.

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ValidationHelper.cs [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+                                if ((!element1.Equals(element2)))
+                                {
+                                    diff.Add(string.Format("Values of Collection {0} does not match.", prop.Name));
+                                }
+                            }
+                            else
+                            {
+                                EvaluateDifferences(element1, element2, diff);
+                            }
+
```
Comment (KanishManuja-MS) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/ValidationHelper.cs]: nit: extra line.

File: test/EndToEndTests/Tests/Client/Build.Desktop/Microsoft.Test.OData.Tests.Client.NetCore.csproj [hunk 5]
```diff
@@ -1,147 +1,147 @@
...
+    <ProjectReference Include="..\..\..\..\..\src\Microsoft.OData.Edm\Microsoft.OData.Edm.NetStandard.VS2017.csproj" />
+    <ProjectReference Include="..\..\..\..\..\src\Microsoft.Spatial\Microsoft.Spatial.NetStandard.VS2017.csproj" />
+    <ProjectReference Include="..\..\..\..\Common\Microsoft.Test.OData.DependencyInjection\Microsoft.Test.OData.DependencyInjection.NetCore.csproj" />
+  </ItemGroup>
+
   <ItemGroup>
     <Folder Include="References\" />
   </ItemGroup>
-
-</Project>
```
Comment (KanishManuja-MS) [inline test/EndToEndTests/Tests/Client/Build.Desktop/Microsoft.Test.OData.Tests.Client.NetCore.csproj]: Does this have any change? If it is a change that you did not intend then please revert this.

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ODataMessageWriterSettingsTests.cs [hunk 6]
```diff
@@ -331,6 +331,49 @@ public void CopyConstructorShouldCopyODataUri()
...
+            this.settings.Version = ODataVersion.V4;
+
+            Func<string, bool> filter = name => true;
+            this.settings.ShouldIncludeAnnotation = filter;
+
+            var newSetting = this.settings.Clone();
+
+            var differences = ValidationHelper.GetDifferences<ODataMessageWriterSettings>(this.settings, newSetting);
+            Assert.True(differences.Count == 0, String.Join(",", differences));
+
```
Comment (KanishManuja-MS) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/ODataMessageWriterSettingsTests.cs]: nit: extra line.

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ODataMessageWriterSettingsTests.cs [hunk 7]
```diff
@@ -357,7 +400,7 @@ public void PropertyGettersAndSettersTest()
                     MaxNestingDepth = maxNestingDepth,
                 },
                 Version = version,
-                LibraryCompatibility = library,
+                LibraryCompatibility = library,
```
Comment (KanishManuja-MS) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/ODataMessageWriterSettingsTests.cs]: revert whitespace.

File: test/FunctionalTests/Microsoft.OData.Core.Tests/TestCharArrayPool.cs [hunk 8]
```diff
@@ -0,0 +1,24 @@
+﻿using Microsoft.OData.Buffers;
+
+namespace Microsoft.OData.Tests
+{
+    public class TestCharArrayPool : ICharArrayPool
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/TestCharArrayPool.cs]: I remember I had a default implementation for ICharArrayPool. Don't I?

File: test/FunctionalTests/Microsoft.OData.Core.Tests/Build.NetFramework/Microsoft.OData.Core.Tests.csproj [hunk 9]
```diff
@@ -281,6 +281,9 @@
     <Compile Include="..\ScenarioTests\UriParser\TopAndSkipFunctionalTests.cs" />
     <Compile Include="..\ScenarioTests\UriParser\VerificationHelpers.cs" />
     <Compile Include="..\ScenarioTests\Writer\JsonLight\WriteFeedWithoutNavigationSourceTests.cs" />
+    <Compile Include="..\TestCharArrayPool.cs">
+      <Link>TestCharArrayPool.cs</Link>
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/Build.NetFramework/Microsoft.OData.Core.Tests.csproj]: Shall we add the <Link> for the Compile element?

File: test/FunctionalTests/Microsoft.OData.Core.Tests/Build.NetFramework/Microsoft.OData.Core.Tests.csproj [hunk 10]
```diff
@@ -450,6 +453,9 @@
     <Compile Include="..\UriParser\Visitors\IsCollectionTranslatorTests.cs" />
     <Compile Include="..\UriParser\Visitors\QueryNodeVisitorTests.cs" />
     <Compile Include="..\UriParser\Visitors\SyntacticTreeVisitorTests.cs" />
+    <Compile Include="..\ValidationHelper.cs">
+      <Link>ValidationHelper.cs</Link>
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/Build.NetFramework/Microsoft.OData.Core.Tests.csproj]: Shall we add the <Link> for the Compile element?

File: test/EndToEndTests/Tests/Client/Build.Desktop/Microsoft.Test.OData.Tests.Client.NetCore.csproj [hunk 11]
```diff
@@ -1,147 +1,147 @@
-<Project Sdk="Microsoft.NET.Sdk">
```
Comment (xuzhg) [inline test/EndToEndTests/Tests/Client/Build.Desktop/Microsoft.Test.OData.Tests.Client.NetCore.csproj]: why change this file?

File: test/FunctionalTests/Microsoft.OData.Core.Tests/TestCharArrayPool.cs [hunk 12]
```diff
@@ -0,0 +1,27 @@
+﻿using Microsoft.OData.Buffers;
+using System;
+
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/TestCharArrayPool.cs]: copyright

File: test/FunctionalTests/Microsoft.OData.Core.Tests/TestCharArrayPool.cs [hunk 13]
```diff
@@ -0,0 +1,27 @@
+﻿using Microsoft.OData.Buffers;
+using System;
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/TestCharArrayPool.cs]: sorting

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ValidationHelper.cs [hunk 14]
```diff
@@ -0,0 +1,117 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/ValidationHelper.cs]: copyright

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ValidationHelper.cs [hunk 15]
```diff
@@ -0,0 +1,117 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.OData.Tests
+{
+    public class ValidationHelper
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/ValidationHelper.cs]: >class ValidationHelper [](start = 11, length = 22)

make it as "static class"?

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ValidationHelper.cs [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+namespace Microsoft.OData.Tests
+{
+    public class ValidationHelper
+    {
+        /// <summary>
+        /// This method takes 2 generic types and see if one is the deep copy  of other
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="obj"></param>
+        /// <param name="copy"></param>
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/ValidationHelper.cs]: If you add the comments, finish the comments

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ValidationHelper.cs [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+            return diff;
+        }
+
+        /// <summary>
+        /// Generic Method to evaluate differences of 2 objects, caled recursively
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="obj"></param>
+        /// <param name="copy"></param>
+        /// <param name="diff"></param>
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/ValidationHelper.cs]: same here?

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ValidationHelper.cs [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+        private static void EvaluateDifferences<T>(T obj, T copy, List<string> diff)
+        {
+
+            foreach (var prop in obj.GetType().GetProperties().Where(x => x.CanRead))
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/ValidationHelper.cs]: verify the input and remove unused the line

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ValidationHelper.cs [hunk 19]
```diff
@@ -0,0 +1,117 @@
...
+                            {
+                                if ((!element1.Equals(element2)))
+                                {
+                                    diff.Add(string.Format("Values of Collection {0} does not match.", prop.Name));
+                                }
+                            }
+                            else
+                            {
+                                EvaluateDifferences(element1, element2, diff);
+                            }
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/ValidationHelper.cs]: It seems you can replace all by calling EvaluateDifferences again?

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ODataMessageReaderSettingsTests.cs [hunk 20]
```diff
@@ -205,6 +205,9 @@ private void CompareMessageReaderSettings(ODataMessageReaderSettings expected, O
             Assert.True(expected.MessageQuotas.MaxOperationsPerChangeset == actual.MessageQuotas.MaxOperationsPerChangeset, "MaxOperationsPerChangeset does not match");
             Assert.True(expected.MessageQuotas.MaxNestingDepth == actual.MessageQuotas.MaxNestingDepth, "MaxNestingDepth does not match");
             Assert.True(expected.MessageQuotas.MaxReceivedMessageSize == actual.MessageQuotas.MaxReceivedMessageSize, "MaxMessageSize does not match");
+
+            var differences = ValidationHelper.GetDifferences<ODataMessageReaderSettings>(expected, actual);
+            Assert.True(differences.Count == 0, String.Join(",", differences));
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/ODataMessageReaderSettingsTests.cs]: If you add the "GetDifferences<T>" function call, how about the codes from Line 191 to Line 207?

File: test/FunctionalTests/Microsoft.OData.Core.Tests/TestCharArrayPool.cs (hunk 13 above)
Comment (Sreejithpin) [reply test/FunctionalTests/Microsoft.OData.Core.Tests/TestCharArrayPool.cs]: REsolved

File: test/FunctionalTests/Microsoft.OData.Core.Tests/TestCharArrayPool.cs (hunk 12 above)
Comment (Sreejithpin) [reply test/FunctionalTests/Microsoft.OData.Core.Tests/TestCharArrayPool.cs]: Resolved

File: test/FunctionalTests/Microsoft.OData.Core.Tests/TestCharArrayPool.cs (hunk 8 above)
Comment (Sreejithpin) [reply test/FunctionalTests/Microsoft.OData.Core.Tests/TestCharArrayPool.cs]: kept the Testchararraypool as such to be reused later as the the testarraypool implementation was inside another class file.

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ValidationHelper.cs (hunk 14 above)
Comment (Sreejithpin) [reply test/FunctionalTests/Microsoft.OData.Core.Tests/ValidationHelper.cs]: REsolved

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ValidationHelper.cs (hunk 15 above)
Comment (Sreejithpin) [reply test/FunctionalTests/Microsoft.OData.Core.Tests/ValidationHelper.cs]: Resolved

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ValidationHelper.cs (hunk 16 above)
Comment (Sreejithpin) [reply test/FunctionalTests/Microsoft.OData.Core.Tests/ValidationHelper.cs]: Resolved

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ValidationHelper.cs (hunk 17 above)
Comment (Sreejithpin) [reply test/FunctionalTests/Microsoft.OData.Core.Tests/ValidationHelper.cs]: Resolved

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ValidationHelper.cs (hunk 18 above)
Comment (Sreejithpin) [reply test/FunctionalTests/Microsoft.OData.Core.Tests/ValidationHelper.cs]: Resolved

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ODataMessageReaderSettingsTests.cs (hunk 20 above)
Comment (Sreejithpin) [reply test/FunctionalTests/Microsoft.OData.Core.Tests/ODataMessageReaderSettingsTests.cs]: Didn't want to change the existing code. Existing lines will function as such

Comment (xuzhg) [reply test/FunctionalTests/Microsoft.OData.Core.Tests/ODataMessageReaderSettingsTests.cs]: That's my concern. It seems we verify the same twice.

---
In reply to: [348177998](https://github.com/OData/odata.net/pull/1595#discussion_r348177998) [](ancestors = 348177998)

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ValidationHelper.cs (hunk 19 above)
Comment (Sreejithpin) [reply test/FunctionalTests/Microsoft.OData.Core.Tests/ValidationHelper.cs]: The text for this contains collection, eg: "Values of Collection {0} does not match.". So to keep that text and any further modificatioins needed for collection did like this. Pls let me know if thats fine

Comment (Sreejithpin) [reply test/FunctionalTests/Microsoft.OData.Core.Tests/ValidationHelper.cs]: Resolved

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ValidationHelper.cs [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+        /// <summary>
+        /// To evaluate the differences of 2 generic types, called recursively
+        /// </summary>
+        /// <typeparam name="T">Method accepts Generic type</typeparam>
+        /// <param name="obj">Object to be compared with</param>
+        /// <param name="copy">Copy object to compare</param>
+        /// <param name="diff">List Of Differences</param>
+        /// <param name="objName">Optional parameter for objectname, to be used in differences string, default empty string</param>
+        /// <param name="isCollection">Optional parameter to see if its a collection, default false</param>
+        private static void EvaluateDifferences<T>(T obj, T copy, List<string> diff,string objName ="", bool isCollection =false)
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/ValidationHelper.cs]: >,string objName ="" [](start = 83, length = 19)

add the whitespace correctly. for example:
add whitespace between the parameter, add the whitespace before and after "="?

File: test/FunctionalTests/Microsoft.OData.Core.Tests/ODataMessageReaderSettingsTests.cs [hunk 22]
```diff
@@ -187,24 +187,9 @@ private void CompareMessageReaderSettings(ODataMessageReaderSettings expected, O
...
-            Assert.True(expected.ShouldIncludeAnnotation == actual.ShouldIncludeAnnotation, "UseKeyAsSegment does not match");
-            Assert.True(expected.Validations == actual.Validations, "Validations does not match");
-            Assert.True(expected.MaxProtocolVersion == actual.MaxProtocolVersion, "MaxProtocolVersion does not match.");
-            Assert.True(expected.MessageQuotas.MaxPartsPerBatch == actual.MessageQuotas.MaxPartsPerBatch, "MaxPartsPerBatch does not match");
-            Assert.True(expected.MessageQuotas.MaxOperationsPerChangeset == actual.MessageQuotas.MaxOperationsPerChangeset, "MaxOperationsPerChangeset does not match");
-            Assert.True(expected.MessageQuotas.MaxNestingDepth == actual.MessageQuotas.MaxNestingDepth, "MaxNestingDepth does not match");
-            Assert.True(expected.MessageQuotas.MaxReceivedMessageSize == actual.MessageQuotas.MaxReceivedMessageSize, "MaxMessageSize does not match");
+            
+            var differences = ValidationHelper.GetDifferences<ODataMessageReaderSettings>(expected, actual);
+            Assert.True(differences.Count == 0, String.Join(",", differences));
```
Comment (xuzhg) [inline test/FunctionalTests/Microsoft.OData.Core.Tests/ODataMessageReaderSettingsTests.cs]: >       [](start = 79, length = 6)

remove the unused whitespace

**Pull request merged**